Master-Slave with Automatic Failover

bradgignac opened the pull request:
**Purpose**

This pull requests implements single master registration with automatic failover to slaves. Additionally, it refactors the existing code and increases test coverage for `app.go` and all registry-related code.

**Implementation**

- Extracts registry-related code in to a new package, `registry`.
- Adds an interface named `Registry` and four implementations:
  - `GroupMasterRegistry` - Allows a single master within a group of services where a group is specified by a unique identifier. Currently, this strategy is used internally but could be useful for implementing master per pod deployments in the future.
  - `SingleMasterRegistry` - Allows one master. This uses `GroupMasterRegistry` by hardcoding the group ID to `master`.
  - `MultiMasterRegistry` - Allows multiple masters. This uses `GroupMasterRegistry` by generating a unique group ID for each service. This strategy follows the same behavior that currently exists in Scroll.
  - `SkipRegistry` - Skips registration with etcd. This implementation is a no-op.
- Moves heartbeating logic out of `scroll.App` and into the `Heartbeater` class in the `registry` package.

Along with the major changes above, I also tried to improve the README, add more test coverage, and remove uses of legacy Vulcan terminology. I also tried to move Mailgun-specific concerns (etcd prefix, polling time, etc) into configuration. Finally, I added an examples directory where we can store examples of how to use Scroll.

**BREAKING CHANGES**

- `scroll.AppConfig` contains a field for Registry in place of the boolean that previously existed.
- This removes support for the legacy Vulcan API.

bradgignac commented on app.go:
```diff
@@ -140,35 +150,20 @@ func (app *App) IsPublicRequest(request *http.Request) bool {
-		}()
-	}
+	// toggle heartbeat on SIGUSR1
+	go func() {
```
I need to re-add logging when heartbeat is paused or restarted.

bradgignac replied:
Fixed.

bradgignac commented on app.go:
```diff
@@ -223,19 +197,16 @@ func (app *App) registerLocationForScope(methods []string, path string, scope Sc
+		Methods:     methods,
+		Middlewares: middlewares,
 	}
+	app.Config.Registry.RegisterHandler(registration)
```
I need to re-add logging here and stop ignoring this error.

bradgignac commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+package registry
```
Need logging in this package.

klizhentas replied:
I would just rename it leader btw to get rid of `master`-`slave` notation

bradgignac commented:
Ping @klizhentas @r0mant.

klizhentas commented on app.go:
```diff
@@ -115,10 +128,7 @@ func (app *App) AddHandler(spec Spec) error {
-	if app.registry != nil && spec.Register != false {
-		app.registerLocation(spec.Methods, spec.Path, spec.Scopes, spec.Middlewares)
-	}
+	app.registerLocation(spec.Methods, spec.Path, spec.Scopes, spec.Middlewares)
```
Does that mean that we always register now? Sometimes I need to turn it off (e.g. when using scoll in Vulcand itself)

bradgignac replied:
With Vulcand, do you turn off registration completely or only for some handlers? You can still choose to disable registration entirely. This does away with per-handler registration because I haven't actually seen it used anywhere.

klizhentas replied:
Completely - Vulcand does not register itself at all

klizhentas commented on app.go:
```diff
@@ -140,35 +150,20 @@ func (app *App) IsPublicRequest(request *http.Request) bool {
+		signal.Notify(heartbeatChan, syscall.SIGUSR1)
+
+		for _ = range heartbeatChan {
+			app.heartbeater.Toggle()
```
Logging here would help, yeah

klizhentas commented on app.go:
```diff
@@ -223,19 +197,16 @@ func (app *App) registerLocationForScope(methods []string, path string, scope Sc
-	if err := app.registry.RegisterFrontend(location); err != nil {
-		log.Errorf("Failed to register a frontend: %v %v", location, err)
-		return
+	registration := &registry.HandlerRegistration{
```
too long of a name for a local var, I would just go for `r`

klizhentas commented on examples/multimaster/main.go:
```diff
@@ -0,0 +1,54 @@
+package main
```
Nice, thanks for the example

klizhentas commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+master instance fails, request handling will automatically failover to a slave
+instance.
+*/
+type GroupMasterRegistry struct {
```
How about `LeaderRegistrar` or something like that

klizhentas replied:
I got it now, I think this `GroupMaster` is a bith thick base as it implements everything that both classes know. I would make it thinner or refactor it as a utility functions instead. E.g. the notion of `Master` is not really needed for the simple case, but you still introduce it because this GroupMaster keeps notion of master inside.

klizhentas commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+// RegisterApp adds a new backend and a single server with Vulcand.
+func (s *GroupMasterRegistry) RegisterApp(registration *AppRegistration) error {
```
How about `AppSpec` or just `App` to shorten it a bit

klizhentas commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+
+// NewGroupMasterRegistry creates a new GroupMasterRegistry from the provided etcd Client.
+func NewGroupMasterRegistry(key string, group string, ttl uint64) *GroupMasterRegistry {
+	client := etcd.NewClient([]string{"http://127.0.0.1:4001"})
```
I would make this as a constant for start

klizhentas commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+	return nil
+}
+
+func (s *GroupMasterRegistry) maintainMasterRole(endpoint *vulcan.Endpoint) error {
```
how about`tryRegister`

klizhentas commented on registry/group_master.go:
```diff
@@ -0,0 +1,170 @@
+	return s.assumeMasterRole(endpoint)
+}
+
+func (s *GroupMasterRegistry) assumeMasterRole(endpoint *vulcan.Endpoint) error {
```
How about `initAsLeader`

klizhentas commented on registry/multi_master.go:
```diff
@@ -0,0 +1,46 @@
+registry uses GroupMasterRegistry by creating a unique group ID for each
+instance of an application.
+*/
+type MultiMasterRegistry struct {
```
I would call it simply a `LBRegistry` as it suits better for load balancing use-case

klizhentas commented on registry/multi_master.go:
```diff
@@ -0,0 +1,46 @@
+		return nil, err
+	}
+
+	innerRegistry := NewGroupMasterRegistry(key, group, ttl)
```
It's also interesting that this registry uses more complex registry to implement itself. I would factor out common logic in the `Client` for registring/deregistring and will roll out additional features in multi-master one

klizhentas commented on registry/registry.go:
```diff
@@ -0,0 +1,71 @@
+
+// Start begins sending heartbeats.
+func (h *Heartbeater) Start() {
+	h.Running = true
```
Is it possible that it will be used concurrently?

bradgignac replied:
It isn't now, but I can add a lock if you'd like.

klizhentas commented on registry/registry.go:
```diff
@@ -0,0 +1,71 @@
+
+func (h *Heartbeater) heartbeat() {
+	for _ = range h.ticker.C {
+		h.registry.RegisterApp(h.registration)
```
I would catch panics and log errors here

klizhentas replied:
Or may be higher level

klizhentas commented on registry/single_master.go:
```diff
@@ -0,0 +1,44 @@
+instance of an application to handle requests. Internally, this registry uses
+a GroupMasterRegistry with a single, constant group ID.
+*/
+type SingleMasterRegistry struct {
```
I would call this `LeaderFollower` registry and implement compare and swap here instead

bradgignac replied:
Don't we want compare and swap in the LB case as well? If we run into some strange scenario where two instances of a service are both trying to register themselves under the same ID, I would rather one accept traffic for one and let the other complain about not being able to register.

klizhentas commented on registry/skip.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// SkipRegistry is an implementation of Registry for applications that do not need service discovery.
+type SkipRegistry struct {
```
`NopRegistry` - as `Go` uses this term in standard lib `NopCloser`

jeremyschlatter commented on README.md:
```diff
@@ -1,12 +1,24 @@
+
+## Getting Started
+
+Build an application with Scroll is simple. Here's a server that listens for GET or POST requests to `http://0.0.0.0:8080/resources/{resourceID}` and echoes back the resource ID provided in the URL.
```
Build -> Building

jeremyschlatter commented on app_test.go:
```diff
@@ -0,0 +1,103 @@
+	Registrations []*registry.HandlerRegistration
+}
+
+func (s *FakeRegistry) RegisterApp(registration *registry.AppRegistration) error {
```
"registration" -> ""

jeremyschlatter commented on app_test.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func (s *FakeRegistry) RegisterHandler(registration *registry.HandlerRegistration) error {
+	s.Registrations = append(s.Registrations, registration)
```
"registration" -> "r"

jeremyschlatter commented on app.go:
```diff
@@ -140,35 +150,20 @@ func (app *App) IsPublicRequest(request *http.Request) bool {
+		heartbeatChan := make(chan os.Signal, 1)
+		signal.Notify(heartbeatChan, syscall.SIGUSR1)
+
+		for _ = range heartbeatChan {
```
Are we standardizing on Go 1.4 now? If so you can use the new syntax here:

```go
for range heartbeatChat {
```

bradgignac replied:
+1

klizhentas replied:
go 1.4.1 is a must now

jeremyschlatter commented on registry/registry.go:
```diff
@@ -0,0 +1,71 @@
+	} else {
+		h.Start()
+	}
+}
```
Calling Toggle repeatedly will leak tickers.

jeremyschlatter replied:
You can `close(h.ticker.C)` in `Stop` to avoid this.

bradgignac replied:
I assumed `Stop` handled this, but:

> Stop does not close the channel, to prevent a read from the channel succeeding incorrectly.

Good catch.

bradgignac replied:
Fixed with a quit channel.

jeremyschlatter commented on registry/registry.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func (h *Heartbeater) heartbeat() {
+	for _ = range h.ticker.C {
```
```go
for range h.ticker.C {
```

jeremyschlatter commented on registry/registry_test.go:
```diff
@@ -0,0 +1,46 @@
+	RegistrationCount int
+}
+
+func (s *FakeRegistry) RegisterApp(registration *AppRegistration) error {
```
"registration" -> ""

The shortest variable name :)

jeremyschlatter commented on registry/registry_test.go:
```diff
@@ -0,0 +1,46 @@
+	return nil
+}
+
+func (s *FakeRegistry) RegisterHandler(registration *HandlerRegistration) error {
```
Same as above.

jeremyschlatter commented on registry/registry_test.go:
```diff
@@ -0,0 +1,46 @@
+	heartbeater.Stop()
+	time.Sleep(30 * time.Millisecond)
+
+	c.Assert(registry.RegistrationCount, Equals, 3)
```
Maybe check that starting it again works too?

bradgignac replied:
Fixed.

jeremyschlatter commented on registry/registry.go:
```diff
@@ -0,0 +1,80 @@
+		case <-h.ticker.C:
+			h.registry.RegisterApp(h.registration)
+		case <-h.quit:
+			break
```
break -> return

`break` will just break out of the select statement, not the for loop.

klizhentas commented:
@bradgignac @jeremyschlatter are we good to merge it?

klizhentas commented:
@bradgignac can you please finish it, I may need it soon for POD service

jeremyschlatter commented:
lgtm

bradgignac commented:
@klizhentas I think I addresses all your comments. The concept of master-slave has been rename to leader-follower. Also, interaction with Vulcan/etcd has been moved out of the registries into a separate package. LBRegistry no longer implements itself using LeaderRegistry. Instead, each one uses the newly extract Vulcan client.

klizhentas commented:
@bradgignac lgtm
